File: 2020/10/week4/minhyungPark/algorithm/Leetcode1626.kt
/**
 * Leetcode 1626. Best Team With No Conflicts
 * https://leetcode.com/contest/weekly-contest-211/problems/best-team-with-no-conflicts/
 */

import kotlin.math.max

class Leetcode1626 {
    fun bestTeamScore(scores: IntArray, ages: IntArray): Int {
        val players = List(scores.size) { ages[it] to scores[it] }.sortedWith(compareBy({it.first}, {it.second})).reversed()
        val dp = IntArray(scores.size)

        for (i in players.indices) {
            val score = players[i].second
            dp[i] = score
            for (j in 0 until i) {
                if (players[j].second >= players[i].second) {
                    dp[i] = max(dp[i], dp[j] + score)
                }
            }
        }
        return dp.max()!!
    }
}

File: 2020/11/week1/1ilsang/algorithm/leetcode1626.js
/**
 * @param {number[]} scores
 * @param {number[]} ages
 * @return {number}
 */
const bestTeamScore = (scores, ages) => {
  const arr = [];
  const dp = [];
  let res = 0;
  for (let i = 0; i < ages.length; i++) {
    arr.push({
      s: scores[i],
      a: ages[i],
    });
  }

  arr.sort((a, b) => {
    if (a.a === b.a) {
      return b.s - a.s;
    }
    return b.a - a.a;
  });

  for (let i = 0; i < arr.length; i++) {
    let max = 0;
    for (let j = 0; j < i; j++) {
      // no conflicts
      if (arr[i].s <= arr[j].s) {
        max = Math.max(max, dp[j]);
      }
    }
    dp[i] = arr[i].s + max;
    res = Math.max(res, dp[i]);
  }
  return res;
};
